### PR TITLE
Add Metric which Keeps Track of Transactions' Memory

### PR DESCRIPTION
**Pull request by bfish713**
#closes: https://github.com/EspressoSystems/HotShot/issues/776

### CODE REVIEW

File: consensus/src/lib.rs [hunk 1]
```diff
@@ -86,6 +86,9 @@ pub struct Consensus<TYPES: NodeTypes> {
     /// Amount of invalid QCs we've seen since the last commit
     /// Used for metrics.  This resets to 0 on every decide event.
     pub invalid_qc: usize,
+
+    /// Used for metrics, the size of the pending transactions in memory.
```
Comment (DieracDelta) [inline consensus/src/lib.rs:90]: Might be worth mentioning here that the size is measured in bytes

**Review by DieracDelta — APPROVED** [review]

File: consensus/src/replica.rs [hunk 2]
```diff
@@ -373,6 +375,15 @@ impl<A: ConsensusApi<TYPES>, TYPES: NodeTypes> Replica<A, TYPES> {
                 .metrics
                 .invalid_qc_views
                 .add_point(consensus.invalid_qc as f64);
+            #[allow(clippy::cast_possible_truncation)]
+            let included_txn_size = included_txns_set.iter().fold(0, |total, txn| {
+                total + bincode_opts().serialized_size(&txn).unwrap_or(0) as usize
+            });
+            consensus.pending_transaction_mem -= included_txn_size;
```
Comment (jbearer) [inline consensus/src/replica.rs:382]: Should we add something like `update(delta: i64)` to the `Gauge` trait so we don't have to store the current value of the gauge separately for metrics like this?

Comment (bfish713) [reply consensus/src/replica.rs:382]: Yeah i've been thinking of this, I'm going to add it.  I thought it had to do with the Prometheus gauge type not having any kind of update but I checked and it expose, get, add and subtract.

File: consensus/src/replica.rs [hunk 3]
```diff
@@ -375,15 +375,14 @@ impl<A: ConsensusApi<TYPES>, TYPES: NodeTypes> Replica<A, TYPES> {
...
-                total + bincode_opts().serialized_size(&txn).unwrap_or(0) as usize
-            });
-            consensus.pending_transaction_mem -= included_txn_size;
+                total + bincode_opts().serialized_size(&txn).unwrap_or(0)
+            }) as i64;
             consensus
                 .metrics
                 .outstanding_transactions_memory_size
-                .set(consensus.pending_transaction_mem);
+                .update(included_txn_size);
```
Comment (jbearer) [inline consensus/src/replica.rs:385]: Should this be `-included_txn_size`?

Comment (bfish713) [reply consensus/src/replica.rs]: 🤦Yes good catch

**Review by jbearer — APPROVED** [review]
LGTM

**Pull request merged**